### DISCUSSION TITLE
How to advance to next month

### DISCUSSION THREAD

**Discussion by tisonkun** [BurntSushi/jiff#131] [Q&A] [answered]
This snippet failed:

```rust
next = advance_time_and_round(next, 1.month(), Some(Unit::Month))?;

fn advance_time_and_round(zoned: Zoned, span: Span, unit: Option<Unit>) -> Result<Zoned, Error> {
    let mut next = zoned;

    next = next
        .checked_add(span)
        .map_err(time_error_with_context(&format!(
            "failed to advance timestamp; end with {next}"
        )))?;

    if let Some(unit) = unit {
        next = next.round(unit).map_err(time_error_with_context(&format!(
            "failed to round timestamp; end with {next}"
        )))?;
    }

    Ok(next)
}
```

```
called `Result::unwrap()` on an `Err` value: Error("failed to round timestamp; end with 2024-10-01T08:01:00+08:00[Asia/Shanghai]: datetime rounding does not support months")
thread 'tests::test_find_next' panicked at src/lib.rs:176:59:
called `Result::unwrap()` on an `Err` value: Error("failed to round timestamp; end with 2024-10-01T08:01:00+08:00[Asia/Shanghai]: datetime rounding does not support months")
stack backtrace:
```

I wonder if there is an alternative to advent to next month correctly.

**Comment by tisonkun** [comment]
Perhaps:

```rust
let rest_days = next.days_in_month() - next.day() + 1;
next = advance_time_and_round(next, rest_days.days(), Some(Unit::Day))?;
```

But I don't know if there is some edge case.

**Comment by BurntSushi** [comment, marked as answer]
For the specific error you're seeing, see #1.

As for how to advance to the next month, just add it: `zdt.checked_add(1.month())`.

It looks like you're doing that already. So it looks like what your code is doing is something more than just "go to next month." Perhaps what you want is to "go to the first day of the next month." I would do that by `zdt.first_of_month()?.checked_add(1.month())?`.

Also, in the future, please provide a complete MRE. Thanks.

**Reply by tisonkun** [reply]
> zdt.first_of_month()?.checked_add(1.month())?

Thank you! This looks a good alternative to try.

> Also, in the future, please provide a complete MRE. Thanks.

OK. Out of curiosity, what is the full form of `MRE`?

**Reply by tisonkun** [reply]
FWIW the full library is now public at https://github.com/tisonkun/cronexpr/blob/main/src/lib.rs

I'm using jiff to build a crontab driver that return the next timestamp based on the given timestamp and the crontab expression.

I plan to finish its usability in days while using it in my database project. This can be part of the jiff ecosystem - just for your information.

**Reply by BurntSushi** [reply]
Nice! 

MRE is "minimal reproducible example." You can search around for instructions on how to provide one. The main idea is that you give a complete reproduction of what you're seeing so that folks helping you don't have to guess.
